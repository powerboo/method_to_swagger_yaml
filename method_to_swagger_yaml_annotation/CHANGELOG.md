## 0.0.3

- add IgnoreFieldInYaml annotation

## 0.0.2

- add useToString option in RequestParameter

## 0.0.1

- init
