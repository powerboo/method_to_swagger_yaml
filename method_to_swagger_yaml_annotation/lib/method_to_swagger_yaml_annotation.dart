library method_to_swagger_yaml_annotation;

class ConvertTargetClass {
  final String title;
  final String? description;
  final String version;
  final List<Tag> tags;

  const ConvertTargetClass({
    required this.title,
    required this.version,
    this.description,
    this.tags = const [],
  });
}

class ConvertTargetMethod {
  final HttpMethodDiv httpMethod;
  final String pathName;
  final List<String> tags;
  final String? summary;
  final String? operationId;

  /// Description for path
  final String? description;

  /// Description for HttpStatusCode 200
  final String responseDescription;

  const ConvertTargetMethod({
    required this.httpMethod,
    required this.pathName,
    this.description,
    this.responseDescription = "Successful operation",
    this.operationId,
    this.tags = const [],
    this.summary,
  });
}

class RequestParameter {
  final RequestParameterDiv requestParameterDiv;
  final String? description;
  final PathParameterStyleDiv? pathStyle;
  final QueryParameterStyleDiv? queryStyle;
  final bool? explode;
  final bool useToString;

  const RequestParameter({
    this.requestParameterDiv = RequestParameterDiv.query,
    this.description,
    this.pathStyle,
    this.queryStyle,
    this.explode,
    this.useToString = false,
  });
}

enum QueryParameterStyleDiv {
  from("from"),
  spaceDelimited("spaceDelimited"),
  pipeDelimited("pipeDelimited"),
  deepObject("deepObject"),
  ;

  final String toStringValue;

  const QueryParameterStyleDiv(this.toStringValue);
  factory QueryParameterStyleDiv.fromString({
    required String value,
  }) {
    final indexEnum = QueryParameterStyleDiv.values
        .firstWhere((e) => e.toStringValue == value);
    return indexEnum;
  }
}

enum PathParameterStyleDiv {
  simple("simple"),
  label("label"),
  matrix("matrix"),
  ;

  final String toStringValue;

  const PathParameterStyleDiv(this.toStringValue);
  factory PathParameterStyleDiv.from({
    required String value,
  }) {
    final indexEnum = PathParameterStyleDiv.values
        .firstWhere((e) => e.toStringValue == value);
    return indexEnum;
  }
}

enum RequestParameterDiv {
  query("query"),
  path("path"),
  /*
  TODO: implements
  header("header"),
  cookie("cookie"),
  // */
  ;

  final String toStringValue;

  const RequestParameterDiv(this.toStringValue);
  factory RequestParameterDiv.from({
    required String value,
  }) {
    final indexEnum =
        RequestParameterDiv.values.firstWhere((e) => e.toStringValue == value);
    return indexEnum;
  }
}

enum HttpMethodDiv {
  get("get"),
  post("post"),
  put("put"),
  patch("patch"),
  delete("delete"),
  ;

  final String toStringValue;

  const HttpMethodDiv(this.toStringValue);
  factory HttpMethodDiv.from({
    required String value,
  }) {
    final indexEnum =
        HttpMethodDiv.values.firstWhere((e) => e.toStringValue == value);
    return indexEnum;
  }
}

class ExternalDocs {
  final String? description;
  final String url;

  const ExternalDocs({
    required this.url,
    this.description,
  });
}

class ExternalDocsException implements Exception {
  late final String message;
  ExternalDocsException(final String message) {
    this.message = "[ExternalDocsException] $message";
  }
  @override
  String toString() {
    return message;
  }
}

class Tag {
  final String name;
  final String? description;
  final ExternalDocs? externalDocs;

  const Tag({
    required this.name,
    this.description,
    this.externalDocs,
  });
}

class TagException implements Exception {
  late final String message;
  TagException(final String message) {
    this.message = "[TagException] $message";
  }
  @override
  String toString() {
    return message;
  }
}
