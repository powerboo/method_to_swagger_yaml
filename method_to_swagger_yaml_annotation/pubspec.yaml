name: method_to_swagger_yaml_annotation
description: "method to swagger yaml annotation."
version: 0.0.1
repository: https://github.com/powerboo/method_to_swagger_yaml

environment:
  sdk: ">=3.2.4 <4.0.0"
