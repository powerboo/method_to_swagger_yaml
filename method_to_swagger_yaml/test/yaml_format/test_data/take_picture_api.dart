import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:method_to_swagger_yaml_annotation/method_to_swagger_yaml_annotation.dart';

import 'value_object/picture_id.dart';

part 'take_picture_api.g.dart';
part 'take_picture_api.freezed.dart';

@ConvertTargetClass(
  title: "title",
  version: "0.0.1",
)
abstract interface class TargetPictureApi {
  @ConvertTargetMethod(
    httpMethod: HttpMethodDiv.get,
    pathName: "target-picture/:pictureId",
  )
  Future<TargetPicture?> find(
      {@RequestParameter(
        requestParameterDiv: RequestParameterDiv.path,
      )
      required PictureId pictureId});

  @ConvertTargetMethod(
    httpMethod: HttpMethodDiv.get,
    pathName: "target-picture",
  )
  Future<List<TargetPicture>> all({
    int cursor = 0,
    int length = 100,
  });

  @ConvertTargetMethod(
    httpMethod: HttpMethodDiv.post,
    pathName: "target-picture",
  )
  Future<void> save({required TargetPicture targetPicture});

  @ConvertTargetMethod(
    httpMethod: HttpMethodDiv.delete,
    pathName: "target-picture/:pictureId",
  )
  Future<void> delete({required PictureId pictureId});
}

@freezed
class TargetPicture with _$TargetPicture {
  const TargetPicture._();
  const factory TargetPicture({
    required PictureId pictureId,
    required PictureName pictureName,
  }) = _TargetPicture;
  factory TargetPicture.fromJson(Map<String, Object?> json) =>
      _$TargetPictureFromJson(json);
}

class TargetPictureException implements Exception {
  late final String message;
  TargetPictureException(final String message) {
    this.message = "[TargetPictureException] $message";
  }
  @override
  String toString() {
    return message;
  }
}

@JsonSerializable()
class PictureId {
  @JsonKey(name: 'picture_id')
  final String value;

  @JsonKey(name: 'picture_kind')
  final PictureKind pictureKind;

  PictureId({
    required this.value,
    required this.pictureKind,
  }) {
    if (value.isEmpty) {
      throw PictureIdException("PictureId does not empty.");
    }
  }

  factory PictureId.fromJson(Map<String, Object?> json) =>
      _$PictureIdFromJson(json);

  Map<String, Object?> toJson() => _$PictureIdToJson(this);

  @override
  bool operator ==(Object other) =>
      identical(other, this) ||
      (other is PictureId &&
          other.value == value &&
          other.pictureKind == pictureKind);

  @override
  int get hashCode => runtimeType.hashCode ^ value.hashCode;
}

@JsonSerializable()
class PictureName {
  @JsonKey(name: 'picture_name')
  final String value;

  PictureName(this.value) {
    if (value.isEmpty) {
      throw Exception("PictureName does not empty.");
    }
  }

  factory PictureName.fromJson(Map<String, Object?> json) =>
      _$PictureNameFromJson(json);

  Map<String, Object?> toJson() => _$PictureNameToJson(this);

  @override
  String toString() {
    return value;
  }

  @override
  bool operator ==(Object other) =>
      identical(other, this) || (other is PictureName && other.value == value);

  @override
  int get hashCode => runtimeType.hashCode ^ value.hashCode;
}

enum PictureKind {
  original,
  thumbnail;
}
