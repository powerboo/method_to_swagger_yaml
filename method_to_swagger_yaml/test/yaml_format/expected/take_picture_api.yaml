openapi: 3.0.0
info: 
  title: title 
  version: 0.0.1 
paths: 
  /target-picture/:pictureId: 
    get: 
      operationId: find 
      parameters: 
        - name: pictureId 
          in: query 
          required: true 
          schema: 
            $ref: '#/components/schemas/FindPictureId' 
      responses: 
        '200': 
          description: Successful operation 
  /target-picture: 
    get: 
      operationId: all 
      parameters: 
        - name: cursor 
          in: query 
          required: true 
          schema: 
            type: integer 
        - name: length 
          in: query 
          required: true 
          schema: 
            type: integer 
      responses: 
        '200': 
          description: Successful operation 
          content: 
            application/json: 
              schema: 
                type: integer 
    post: 
      operationId: save 
      responses: 
        '200': 
          description: Successful operation 
components: 
  schemas: 
    FindPictureId: 
      type: object 
      properties: 
        pictureId: 
          type: object 
          properties: 
            value: 
              type: string 
            pictureKind: 
              type: integer 
              enum: [0, 1] 
    AllCursor: 
      type: integer 
    AllLength: 
      type: integer 
